In [148]:
import csv

import pandas as pd
import numpy as np

In [149]:
from scipy.io.arff import loadarff

In [150]:
# Output dataframe with normalized values
CSV_FILE = "wines_dataframe.csv"

In [151]:
df = pd.read_csv("raw_dataframe.csv", encoding="cp1252")
df.head(10)

name    price  rating  \
0  Vino Tracer Riesling, Weinkellerei Hechtsheim,...   1312.0     4.7   
1                     Vino Lighea, Donnafugata, 2021   4990.0     4.8   
2             Vino Chenin Blanc, David & Nadia, 2022   6790.0     4.5   
3  Vino Pinot Noir Alpine Vineyard, Rhys Vineyard...  34990.0     5.0   
4                   Vino Grain de Gris, Listel, 2022   1393.0     4.6   
5  Vino Volnay Premier Cru Clos des Chenes, Domai...  23490.0     5.0   
6  Vino Tenuta Regaleali Cygnus, Tasca d'Almerita...   4490.0     4.5   
7  Vino Bordeaux des Iles, Domaine de l'Ile Marga...   5490.0     5.0   
8  Vino Riesling Kastelberg Grand Cru Le Chateau,...  17990.0     5.0   
9  Vino Primofiore, Giuseppe Quintarelli, 2021, 1...  39990.0     5.0   

                      country              region sweetness  \
0                   Germanija             Pfal'ts  semi-dry   
1                     Italija           Sitsilija       dry   
2             Juzhnaja Afrika           Svortlend       dry   
3  Soedinennye Shtaty Ameriki         Kalifornija       dry   
4                   Frantsija  Langedok-Russil'on       dry   
5                   Frantsija          Burgundija       dry   
6                     Italija           Sitsilija       dry   
7                   Frantsija               Bordo       dry   
8                   Frantsija              El'zas  semi-dry   
9                     Italija              Veneto       dry   

                  grape               manufacturer  strength  volume  
0          risling 100%    Weinkellerei Hechtsheim      12.0    0.75  
1          zibibbo 100%                Donnafugata      12.5    0.75  
2      shenen blan 100%              David & Nadia      12.5    0.75  
3             pino nuar             Rhys Vineyards      12.9    0.75  
4                  sira                     Listel      12.0    0.75  
5        pino nuar 100%      Domaine Antonin Guyon      14.0    0.75  
6  kaberne sovin'on 40%           Tasca d'Almerita      14.0    0.75  
7             pti verdo   Domaine de l'Ile Margaux      13.5    0.75  
8          risling 100%  Domaine Marc Kreydenweiss      13.0    0.75  
9          kaberne fran       Giuseppe Quintarelli      14.0    1.50

In [152]:
# df["price"].argmax()
df.loc[899] # НУИНУ 2 ляма за бутылку

name            Vino Sassicaia, Tenuta San Guido, 2018, 6 l
price                                             1999990.0
rating                                                  5.0
country                                             Italija
region                                              Toskana
sweetness                                               dry
grape                                      kaberne fran 15%
manufacturer                               Tenuta San Guido
strength                                               13.5
volume                                                  6.0
Name: 899, dtype: object

In [153]:
df.describe()

price       rating     strength       volume
count  3.501000e+03  3550.000000  3535.000000  3549.000000
mean   1.861549e+04     3.777155    13.320249     0.787524
std    7.429743e+04     1.790866     1.286435     0.282318
min    5.700000e+02     0.000000     0.500000     0.187000
25%    2.640000e+03     4.100000    12.500000     0.750000
50%    5.018000e+03     4.600000    13.500000     0.750000
75%    1.299000e+04     4.800000    14.000000     0.750000
max    1.999990e+06     5.000000    21.000000     6.000000

In [154]:
df.nunique()

name            3539
price            369
rating            22
country           19
region            78
sweetness          4
grape            422
manufacturer     431
strength          60
volume             9
dtype: int64

In [155]:
nans = df.isnull().any(axis=1).sum()
print(nans)
print(nans / len(df))

112
0.031549295774647886


Можно заметить, что у нас есть $112$ объектов, в которых пропущены какие-то значения. Это всего $3\%$ от общего количества объектов, поэтому я думаю, что можно безопасно от них избавиться

In [156]:
df.dropna(inplace=True)

Теперь нормализуем значения в колонках __price__, __rating__, __strength__ и __volume__

In [157]:
for column in ["price", "strength", "volume"]:
    df[column] = (df[column] - df[column].mean()) / df[column].std()

df.head(10)

name     price  rating  \
0  Vino Tracer Riesling, Weinkellerei Hechtsheim,... -0.229692     4.7   
1                     Vino Lighea, Donnafugata, 2021 -0.180184     4.8   
2             Vino Chenin Blanc, David & Nadia, 2022 -0.155955     4.5   
3  Vino Pinot Noir Alpine Vineyard, Rhys Vineyard...  0.223636     5.0   
4                   Vino Grain de Gris, Listel, 2022 -0.228602     4.6   
5  Vino Volnay Premier Cru Clos des Chenes, Domai...  0.068839     5.0   
6  Vino Tenuta Regaleali Cygnus, Tasca d'Almerita... -0.186914     4.5   
7  Vino Bordeaux des Iles, Domaine de l'Ile Marga... -0.173454     5.0   
8  Vino Riesling Kastelberg Grand Cru Le Chateau,... -0.005195     5.0   
9  Vino Primofiore, Giuseppe Quintarelli, 2021, 1...  0.290940     5.0   

                      country              region sweetness  \
0                   Germanija             Pfal'ts  semi-dry   
1                     Italija           Sitsilija       dry   
2             Juzhnaja Afrika           Svortlend       dry   
3  Soedinennye Shtaty Ameriki         Kalifornija       dry   
4                   Frantsija  Langedok-Russil'on       dry   
5                   Frantsija          Burgundija       dry   
6                     Italija           Sitsilija       dry   
7                   Frantsija               Bordo       dry   
8                   Frantsija              El'zas  semi-dry   
9                     Italija              Veneto       dry   

                  grape               manufacturer  strength    volume  
0          risling 100%    Weinkellerei Hechtsheim -1.133567 -0.121366  
1          zibibbo 100%                Donnafugata -0.695962 -0.121366  
2      shenen blan 100%              David & Nadia -0.695962 -0.121366  
3             pino nuar             Rhys Vineyards -0.345878 -0.121366  
4                  sira                     Listel -1.133567 -0.121366  
5        pino nuar 100%      Domaine Antonin Guyon  0.616852 -0.121366  
6  kaberne sovin'on 40%           Tasca d'Almerita  0.616852 -0.121366  
7             pti verdo   Domaine de l'Ile Margaux  0.179247 -0.121366  
8          risling 100%  Domaine Marc Kreydenweiss -0.258357 -0.121366  
9          kaberne fran       Giuseppe Quintarelli  0.616852  2.631291

Теперь разберемся с категориальными признаками - __country__, __region__, __grape__, __manufacturer__

Начнем с препроцессинга видов винограда



In [158]:
df["grape"]

0             risling 100%
1             zibibbo 100%
2         shenen blan 100%
3                pino nuar
4                     sira
               ...        
3545         shardone 100%
3546          risling 100%
3547    sovin'on blan 100%
3548         sovin'on blan
3549          risling 100%
Name: grape, Length: 3438, dtype: object

In [159]:
def process_grapes(grape: str) -> str:
    if grape.endswith("%"):
        return ' '.join(grape.split()[:-1])
    
    return grape


df["grape"] = df["grape"].map(process_grapes)
df.head()

name     price  rating  \
0  Vino Tracer Riesling, Weinkellerei Hechtsheim,... -0.229692     4.7   
1                     Vino Lighea, Donnafugata, 2021 -0.180184     4.8   
2             Vino Chenin Blanc, David & Nadia, 2022 -0.155955     4.5   
3  Vino Pinot Noir Alpine Vineyard, Rhys Vineyard...  0.223636     5.0   
4                   Vino Grain de Gris, Listel, 2022 -0.228602     4.6   

                      country              region sweetness        grape  \
0                   Germanija             Pfal'ts  semi-dry      risling   
1                     Italija           Sitsilija       dry      zibibbo   
2             Juzhnaja Afrika           Svortlend       dry  shenen blan   
3  Soedinennye Shtaty Ameriki         Kalifornija       dry    pino nuar   
4                   Frantsija  Langedok-Russil'on       dry         sira   

              manufacturer  strength    volume  
0  Weinkellerei Hechtsheim -1.133567 -0.121366  
1              Donnafugata -0.695962 -0.121366  
2            David & Nadia -0.695962 -0.121366  
3           Rhys Vineyards -0.345878 -0.121366  
4                   Listel -1.133567 -0.121366

In [160]:
df = pd.get_dummies(df, columns=["grape"])
df.head()

name     price  rating  \
0  Vino Tracer Riesling, Weinkellerei Hechtsheim,... -0.229692     4.7   
1                     Vino Lighea, Donnafugata, 2021 -0.180184     4.8   
2             Vino Chenin Blanc, David & Nadia, 2022 -0.155955     4.5   
3  Vino Pinot Noir Alpine Vineyard, Rhys Vineyard...  0.223636     5.0   
4                   Vino Grain de Gris, Listel, 2022 -0.228602     4.6   

                      country              region sweetness  \
0                   Germanija             Pfal'ts  semi-dry   
1                     Italija           Sitsilija       dry   
2             Juzhnaja Afrika           Svortlend       dry   
3  Soedinennye Shtaty Ameriki         Kalifornija       dry   
4                   Frantsija  Langedok-Russil'on       dry   

              manufacturer  strength    volume  grape_agiorgitiko  ...  \
0  Weinkellerei Hechtsheim -1.133567 -0.121366              False  ...   
1              Donnafugata -0.695962 -0.121366              False  ...   
2            David & Nadia -0.695962 -0.121366              False  ...   
3           Rhys Vineyards -0.345878 -0.121366              False  ...   
4                   Listel -1.133567 -0.121366              False  ...   

   grape_vermentino  grape_vernachcha  grape_vespolina  grape_vion'e  \
0             False             False            False         False   
1             False             False            False         False   
2             False             False            False         False   
3             False             False            False         False   
4             False             False            False         False   

   grape_viozin'o  grape_vitovska  grape_viura  grape_voskeat  grape_zibibbo  \
0           False           False        False          False          False   
1           False           False        False          False           True   
2           False           False        False          False          False   
3           False           False        False          False          False   
4           False           False        False          False          False   

   grape_zinfandel'  
0             False  
1             False  
2             False  
3             False  
4             False  

[5 rows x 205 columns]

In [161]:
df = pd.get_dummies(df, columns=["sweetness", "country", "region", "manufacturer"])
df.head()

name     price  rating  \
0  Vino Tracer Riesling, Weinkellerei Hechtsheim,... -0.229692     4.7   
1                     Vino Lighea, Donnafugata, 2021 -0.180184     4.8   
2             Vino Chenin Blanc, David & Nadia, 2022 -0.155955     4.5   
3  Vino Pinot Noir Alpine Vineyard, Rhys Vineyard...  0.223636     5.0   
4                   Vino Grain de Gris, Listel, 2022 -0.228602     4.6   

   strength    volume  grape_agiorgitiko  grape_ajren  grape_al'barin'o  \
0 -1.133567 -0.121366              False        False             False   
1 -0.695962 -0.121366              False        False             False   
2 -0.695962 -0.121366              False        False             False   
3 -0.345878 -0.121366              False        False             False   
4 -1.133567 -0.121366              False        False             False   

   grape_al'bil'o  grape_al'janiko  ...  manufacturer_Viticola Mentridana  \
0           False            False  ...                             False   
1           False            False  ...                             False   
2           False            False  ...                             False   
3           False            False  ...                             False   
4           False            False  ...                             False   

   manufacturer_Vodopivec  manufacturer_Volpe Pasini  \
0                   False                      False   
1                   False                      False   
2                   False                      False   
3                   False                      False   
4                   False                      False   

   manufacturer_Weingut Julian Haart  manufacturer_Weinkellerei Hechtsheim  \
0                              False                                  True   
1                              False                                 False   
2                              False                                 False   
3                              False                                 False   
4                              False                                 False   

   manufacturer_William Fevre  manufacturer_Winzer Krems  \
0                       False                      False   
1                       False                      False   
2                       False                      False   
3                       False                      False   
4                       False                      False   

   manufacturer_Yves Cuilleron  manufacturer_Zenato  manufacturer_Zulal  
0                        False                False               False  
1                        False                False               False  
2                        False                False               False  
3                        False                False               False  
4                        False                False               False  

[5 rows x 723 columns]

In [162]:
df.columns

Index(['name', 'price', 'rating', 'strength', 'volume', 'grape_agiorgitiko',
       'grape_ajren', 'grape_al'barin'o', 'grape_al'bil'o', 'grape_al'janiko',
       ...
       'manufacturer_Viticola Mentridana', 'manufacturer_Vodopivec',
       'manufacturer_Volpe Pasini', 'manufacturer_Weingut Julian Haart',
       'manufacturer_Weinkellerei Hechtsheim', 'manufacturer_William Fevre',
       'manufacturer_Winzer Krems', 'manufacturer_Yves Cuilleron',
       'manufacturer_Zenato', 'manufacturer_Zulal'],
      dtype='object', length=723)

In [163]:
print(sorted(df["rating"].unique()))

[np.float64(0.0), np.float64(1.0), np.float64(2.7), np.float64(3.0), np.float64(3.2), np.float64(3.4), np.float64(3.5), np.float64(3.6), np.float64(3.7), np.float64(3.8), np.float64(3.9), np.float64(4.0), np.float64(4.1), np.float64(4.2), np.float64(4.3), np.float64(4.4), np.float64(4.5), np.float64(4.6), np.float64(4.7), np.float64(4.8), np.float64(4.9), np.float64(5.0)]


В качестве таргетного категориального признака я выберу значение рейтинга, предварительно его дискретизировав.

Можно предположить, что есть какие-то выбросы в виде объектов с рейтингом $< 3$ (или хотя бы $\le 1$)

In [164]:
df[df["rating"] < 3].shape
# Можно увидеть, что 593 объекта это все таки не так мало

(593, 723)

In [165]:
df[df["rating"] <= 1].shape

(591, 723)

In [166]:
df[df["rating"] == 0.0].shape

(589, 723)

Попробуем дискретизировать значение рейтинга, переведя его в значения $0 \dots 5$

In [167]:
def get_rating_range(rating: float) -> int:
    ranges = [
        (0, 1),
        (1, 2),
        (2, 3),
        (3, 4),
        (4, 5)
    ]
    
    for idx, rang in enumerate(ranges):
        if rang[0] <= rating < rang[1]:
            return idx + 1
    
    return 5 # last range is closed [4, 5] but others are half-open ( [0, 1) for example)
    

In [168]:
df["rating"] = df["rating"].map(get_rating_range)
df.tail(10)

name     price  rating  \
3540        Vino Primofiore, Giuseppe Quintarelli, 2021 -0.025386       5   
3541                   Vino Villa Solais, Santadi, 2019 -0.209798       4   
3542                        Vino Chateau Ferriere, 2020 -0.079229       5   
3543                       Vino Chateau La Lagune, 2017 -0.079229       5   
3544  Vino Gevrey-Chambertin Clos du Couvent, Domain...  0.001535       1   
3545  Vino Bourgogne Kimmeridgien, Jean-Marc Brocard... -0.189607       5   
3546            Vino Riesling Kalkmergel, Knipser, 2021 -0.173454       5   
3547  Vino Pouilly-Fume La Demoiselle de Bourgeois, ... -0.119611       5   
3548  Vino Semisam Shardone/Sovin'on Blan, Shumrinka... -0.234027       5   
3549  Vino Riesling Ried Loibenberg Smaragd, Emmeric... -0.059038       5   

      strength    volume  grape_agiorgitiko  grape_ajren  grape_al'barin'o  \
3540  0.616852 -0.121366              False        False             False   
3541 -0.258357 -0.121366              False        False             False   
3542  0.179247 -0.121366              False        False             False   
3543  0.179247 -0.121366              False        False             False   
3544 -0.258357 -0.121366              False        False             False   
3545 -0.258357 -0.121366              False        False             False   
3546 -0.695962 -0.121366              False        False             False   
3547  0.616852 -0.121366              False        False             False   
3548 -0.433399 -0.121366              False        False             False   
3549 -0.258357 -0.121366              False        False             False   

      grape_al'bil'o  grape_al'janiko  ...  manufacturer_Viticola Mentridana  \
3540           False            False  ...                             False   
3541           False            False  ...                             False   
3542           False            False  ...                             False   
3543           False            False  ...                             False   
3544           False            False  ...                             False   
3545           False            False  ...                             False   
3546           False            False  ...                             False   
3547           False            False  ...                             False   
3548           False            False  ...                             False   
3549           False            False  ...                             False   

      manufacturer_Vodopivec  manufacturer_Volpe Pasini  \
3540                   False                      False   
3541                   False                      False   
3542                   False                      False   
3543                   False                      False   
3544                   False                      False   
3545                   False                      False   
3546                   False                      False   
3547                   False                      False   
3548                   False                      False   
3549                   False                      False   

      manufacturer_Weingut Julian Haart  manufacturer_Weinkellerei Hechtsheim  \
3540                              False                                 False   
3541                              False                                 False   
3542                              False                                 False   
3543                              False                                 False   
3544                              False                                 False   
3545                              False                                 False   
3546                              False                                 False   
3547                              False                                 False   
3548                              False                                 False   
354

In [169]:
df.to_csv(CSV_FILE, index=False, encoding="cp1252")